ml/kmeans/KMeans.cc

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "KMeans.h"
+#include <dlib/clustering.h>
+
+void KMeans::train(SamplesBuffer &SB) {
+    std::vector<DSample> Samples = SB.preprocess();
+
+    MinDist = std::numeric_limits<CalculatedNumber>::max();
+    MaxDist = std::numeric_limits<CalculatedNumber>::min();
+
+    {
+        std::lock_guard<std::mutex> Lock(Mutex);
+
+        ClusterCenters.clear();
+
+        dlib::pick_initial_centers(NumClusters, ClusterCenters, Samples);
+        dlib::find_clusters_using_kmeans(Samples, ClusterCenters);
```

Comment:
as per Slack discussion, I would think about adding 3rd parameter as configurable thing:
```        >
    void find_clusters_using_kmeans (
        const array_type& samples,
        std::vector<sample_type, alloc>& centers,
        unsigned long max_iter = 1000
    );```
Also do we have any worst case benchmark (in case it really does 1000 iterations)?

Reply:
> Also do we have any worst case benchmark (in case it really does 1000 iterations)?

I'll run a `TrainEvery = 6h` version that tracks the maximum number of iterations to find out. I'd be surprised if it is even close to the limit because we run the function that picks suitable initial clusters first.

> Under which conditions dlib kmeans algorithm stops (other than max_iter)? Is it when no point changes cluster? Or if centroids do not move by more than X?

https://github.com/davisking/dlib/blob/master/dlib/svm/kkmeans.h#L436

Reply:
> I would think about adding 3rd parameter as configurable thing

No objection to this, but I'd prefer to release the MVP first and wait to see how well it performs with the defaults and how easily our users can understand the options that we currently have.

As far as performance is concerned, time to train, train every and lag_n have *a much more* significant impact on performance.

Reply:
Even in case of not making this configurable should we give more strict limit? I think in absolute majority of cases it will stop far before 1000 iterations. But I wonder what is the expense of this in worst case. Do we know how many iterations it takes typically?

Reply:
You are raising a very good point. The problem is that dlib is not returning info on the number of iterations done. Also, thinking about this, I don't see how me building an agent with `TrainN = 6h` is going to help because I'd expect the number of iterations to be different on anomalous servers.

I think we have 3 options:

1. Add this as a configuration option,
2. Modify dlib's condition in the main KMeans loop to force running for 1000 iterations and measure execution time,
3. Surface this information in the ML charts and what is the max value in the instance that Andy will setup on staging.

The 3rd option is problematic because dlib is not returning the number of iterations done. The 2nd option does not allow us to know how _often_ we have to go through many iterations in a live setting. The 1st option would provide someone yet one more configuration option that they could potentially optimize.

To be honest, I think we need a combination of 3 and 1. My suggestion would be to go with 1 for this PR because so far on staging we haven't noticed any performance problems caused by the training step itself. What do you think?

Reply:
FWIW, I forgot to mention:

I tried generating random data sampled from many different distributions. I tested different number of samples, with many dimensions and a big lag_n option. The worst I got was ~180 iterations (for very big number of samples, high dimensionality and high lag_n). Normally, it would be around 60-80 iterations. Most definitely its the call to pick the initial centers that keeps the number of iterations on the low end.

Completely unscientific but it should give an idea on the importance of picking appropriate initial centers.

Reply:
yep - and i think `k` or num clusters will also impact this and i have seen a few examples of dimensions where k=2 leads to a bad model as there is just too much going on in the dimension and probably would get a better model with k=3 for example but that would also then impact any benchmarks and things like this.

If we could just measure average train time (or a histogram ideally) or some sort of high level indication of training cost over time then that could be a solid yet flexible way to measure various optimisations we might try make in future. So i do think there could be some sort of "debug" mode we could build for the ML that maybe exposes more things like this that we can use to measure technical considerations as we develop. But again, probably something for a later stage once we have a much better feel for it and know how best to build this sort of debug mode.

Reply:
> But again, probably something for a later stage once we have a much better feel for it and know how best to build this sort of debug mode.

This is actually something that we need to land with this PR. Otherwise, we won't have any idea what is going on with the systems of our users. I plan doing this (along with adding a couple of log messages when we hit some special conditions in the code) once the majority of the PR's architecture has been reviewed sufficiently, ie. early next week.